ddtrace/tracer/sampler.go

```diff
@@ -147,3 +154,273 @@ func (ps *prioritySampler) apply(spn *span) {
+		// short path when disabled
+		return false
+	}
+
+	var matched bool
+	rate := rs.globalRate
+	for _, rule := range rs.rules {
+		if rule.match(span) {
+			matched = true
+			rate = rule.Rate
+			break
+		}
+	}
+	if !matched && math.IsNaN(rate) {
+		// no matching rule or global rate, so we want to fall back
+		// to priority sampling
+		return false
+	}
+
+	rs.applyRate(span, rate, time.Now())
+	return true
+}
+
+func (rs *rulesSampler) applyRate(span *span, rate float64, now time.Time) {
+	span.SetTag(keyRulesSamplerAppliedRate, rate)
+	if !sampledByRate(span.TraceID, rate) {
+		span.SetTag(ext.SamplingPriority, ext.PriorityAutoReject)
+		return
+	}
+
+	sampled, rate := rs.limiter.samplingDecision(now)
+	if sampled {
+		span.SetTag(ext.SamplingPriority, ext.PriorityAutoKeep)
+	} else {
+		span.SetTag(ext.SamplingPriority, ext.PriorityAutoReject)
+	}
+	span.SetTag(keyRulesSamplerLimiterRate, rate)
+}
+
+// SamplingRule is used for applying sampling rates to spans that match
+// the service name, operation name or both.
+// For basic usage, consider using the helper functions ServiceRule, NameRule, etc.
+type SamplingRule struct {
+	Service *regexp.Regexp
+	Name    *regexp.Regexp
+	Rate    float64
+
+	exactService string
+	exactName    string
+}
+
+// ServiceRule returns a SamplingRule that applies the provided sampling rate
+// to spans that match the service name provided.
+func ServiceRule(service string, rate float64) SamplingRule {
+	return SamplingRule{
+		exactService: service,
+		Rate:         rate,
+	}
+}
+
+// NameRule returns a SamplingRule that applies the provided sampling rate
+// to spans that match the operation name provided.
+func NameRule(name string, rate float64) SamplingRule {
+	return SamplingRule{
+		exactName: name,
+		Rate:      rate,
+	}
+}
+
+// NameServiceRule returns a SamplingRule that applies the provided sampling rate
+// to spans matching both the operation and service names provided.
+func NameServiceRule(name string, service string, rate float64) SamplingRule {
+	return SamplingRule{
+		exactService: service,
+		exactName:    name,
+		Rate:         rate,
+	}
+}
+
+// RateRule returns a SamplingRule that applies the provided sampling rate to all spans.
+func RateRule(rate float64) SamplingRule {
+	return SamplingRule{
+		Rate: rate,
+	}
+}
+
+// match returns true when the span's details match all the expected values in the rule.
+func (sr *SamplingRule) match(s *span) bool {
+	if sr.Service != nil && !sr.Service.MatchString(s.Service) {
+		return false
+	} else if sr.exactService != "" && sr.exactService != s.Service {
+		return false
+	}
+	if sr.Name != nil && !sr.Name.MatchString(s.Name) {
+		return false
+	} else if sr.exactName != "" && sr.exactName != s.Name {
+		return false
+	}
+	return true
+}
+
+// samplingLimiter is used to limit the quantity of spans sampled by the rules sampler
+// and internally maintains measurements for "effective rate".
+type samplingLimiter struct {
```

Comment:
This is a rate limiter, not a sampling limiter, can we call it that? Rule of thumb for both naming things and documentation is to describe what they are, not where they are used. e.g.
```
// rateLimiter is a wrapper on top of golang.org/x/time/rate which also keeps track of the effective throughput rate
// within a 2 second window.
type rateLimiter struct {
```